File: backend/src/script/lib/agn3/dbconfig.py
####################################################################################################################################################################################################################################################################
#                                                                                                                                                                                                                                                                  #
#                                                                                                                                                                                                                                                                  #
#        Copyright (C) 2025 AGNITAS AG (https://www.agnitas.org)                                                                                                                                                                                                   #
#                                                                                                                                                                                                                                                                  #
#        This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.    #
#        This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.           #
#        You should have received a copy of the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.                                                                                                            #
#                                                                                                                                                                                                                                                                  #
####################################################################################################################################################################################################################################################################
#
from	__future__ import annotations
import	os, re, json
from	typing import Any, Final, Optional
from	typing import Dict, Iterator, Tuple
from	typing import overload
from	.definitions import base, syscfg, dbid_default
#
class DBConfig:
	"""Database configuration handler

reads and stores configuration from a configuration file"""
	__slots__ = ['path', 'data', 'dbid_default']
	config_env: Final[str] = 'DBCFG'
	config_path_env: Final[str] = 'DBCFG_PATH'
	default_config_path: Final[str] = '/opt/agnitas.com/etc/dbcfg'
	def __init__ (self, path: Optional[str] = None) -> None:
		self.path = self._findpath (path)
		self.data: Dict[str, DBConfig.DBRecord] = {}
		self.dbid_default = dbid_default
		self.read ()
		if len (self.data) == 1:
			self.dbid_default = list (self.data.keys ())[0]
	
	def _findpath (self, path: Optional[str]) -> str:
		if path is not None:
			return path
		if (path := os.environ.get (self.config_path_env)) is not None:
			return path
		path = os.path.join (base, 'etc', 'dbcfg')
		if os.path.isfile (path):
			return path
		return self.default_config_path

	class DBRecord:
		"""A Record for one database configuration line"""
		__slots__ = ['dbid', 'data']
		def __init__ (self, dbid: None | str, param: Optional[str] = None) -> None:
			self.dbid = dbid
			self.data: Dict[str, str] = {}
			if param is not None:
				for elem in [_p.strip () for _p in param.split (', ')]:
					elem = elem.strip ()
					parts = [_e.strip () for _e in elem.split ('=', 1)]
					if len (parts) == 2:
						self.data[parts[0]] = parts[1]
		
		def __str__ (self) -> str:
			return f'{self.__class__.__name__} (dbid = {self.dbid!r}, data = {self.data!r}'
		__repr__ = __str__

		def __getitem__ (self, id: str) -> str:
			return self.data[id]
		
		def __setitem__ (self, id: str, value: str) -> None:
			self.data[id] = value

		def __contains__ (self, id: str) -> bool:
			return id in self.data

		@overload
		def __call__ (self, id: str, default: None = ...) -> None | str: ...
		@overload
		def __call__ (self, id: str, default: str) -> str: ...
		def __call__ (self, id: str, default: None | str = None) -> Any:
			try:
				return self[id]
			except KeyError:
				return default
		
		def update (self, source: Dict[str, str]) -> None:
			self.data.update (source)
		
		def copy (self) -> DBConfig.DBRecord:
			rc = self.__class__ (self.dbid, None)
			rc.update (self.data)
			return rc

	parse_line = re.compile ('([a-z0-9._+-]+):[ \t]*(.*)$', re.IGNORECASE)
	def read (self) -> None:
		self.data.clear ()
		content: None | str = None
		if (dbcfg := os.environ.get (self.config_env)) is not None:
			try:
				parsed = json.loads (dbcfg)
				if isinstance (parsed, dict):
					jid: str
					jparam: str | dict[str, str]
					for (jid, jparam) in parsed.items ():
						self.data[jid] = record = DBConfig.DBRecord (jid, jparam if isinstance (jparam, str) else None)
						if isinstance (jparam, dict):
							record.update (jparam)
			except Exception:
				content = dbcfg
		elif os.path.isfile (self.path):
			with open (self.path) as fd:
				content = fd.read ()
		else:
			if (dbcfg := syscfg.get ('dbcfg')) is not None:
				self.data[self.dbid_default] = DBConfig.DBRecord (self.dbid_default, dbcfg)
			return
		#
		if content is not None:
			for line in (_l.strip () for _l in content.split ('\n')):
				if line and not line.startswith ('#') and (mtch := self.parse_line.match (line)) is not None:
					(id, param) = mtch.groups ()
					self.data[id] = DBConfig.DBRecord (id, param)

	def __getitem__ (self, id: Optional[str]) -> DBConfig.DBRecord:
		return self.data[id if id is not None else self.dbid_default]

	def __iter__ (self) -> Iterator[str]:
		return iter (self.data)
	
	def keys (self) -> Iterator[str]:
		return iter (self.data.keys ())
	
	def values (self) -> Iterator[DBConfig.DBRecord]:
		return iter (self.data.values ())

	def items (self) -> Iterator[Tuple[str, DBConfig.DBRecord]]:
		return iter (self.data.items ())

	def inject (self, id: str, param: Dict[str, str]) -> None:
		"""adds a configuration which is not part of the configuration file

with this method it is possible to add a configuration for a database
which is not part of the configuration file."""
		rec = DBConfig.DBRecord (id, None)
		rec.data = param
		self.data[id] = rec
	
	def remove (self, id: str) -> None:
		"""remove a configuration entry"""
		if id in self.data:
			del self.data[id]

File: frontend/src/java/com/agnitas/util/DBConfig.java
/*

    Copyright (C) 2025 AGNITAS AG (https://www.agnitas.org)

    This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
    You should have received a copy of the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.

*/

package com.agnitas.util;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.util.HashMap;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * This class reads the global database access configuration file and
 * parses it and provides methods to retrieve these values.
 */
public class DBConfig {
	private static final String	defaultConfigPath = "/opt/agnitas.com/etc/dbcfg";
	private String			configPath = null;
	private HashMap <String, HashMap <String, String>>	
					content = null;
	private HashMap <String, String>
					selected = null;
	private String			lastError = null;

	/**
	 * If an error had occured, return the last error
	 * 
	 * @return the last error occured or null if no error had occured
	 */
	public String getLastError () {
		return lastError;
	}
	
	/**
	 * Set the path to the configuration file. If this had not been set
	 * the default configuration file will be used.
	 * 
	 * @param path the new path to the configuration file.
	 */
	public void setConfigPath (String path) {
		configPath = path;
		content = null;
	}

	/**
	 * Resets the path of the configuration file to its default location
	 */
	public void resetConfigPath () {
		setConfigPath (null);
	}
	
	/**
	 * Search and return the configuration for the provided database id.
	 * 
	 * All configuration are a key/value pair stored in a map which will
	 * be returned. The available entries depend on the configuration and
	 * from the used dbms.
	 * 
	 * @param id the database id for which the record should be retrieved
	 * @return   the record for the id, if found, null otherwise
	 */
	public HashMap <String, String> find (String id) {
		readConfig ();
		return content != null ? content.get (id) : null;
	}
	
	/**
	 * Returns a list of all known database ids from this configuration
	 * 
	 * @return a list (array of strings) of all known database ids or null, if the configuration file is not accessable
	 */
	public String[] ids () {
		String[]	rc = null;
		
		readConfig ();
		if (content != null) {
			Set <String>	keys = content.keySet ();
			
			rc = keys.toArray (new String[0]);
		}
		return rc;
	}
	
	/**
	 * Select a dabatase id for further access the entries using the findInRecord methods
	 * 
	 * @param id the database id to look for
	 * @return   true if the record had been found, false otherwise
	 */
	public boolean selectRecord (String id) {
		selected = find (id);
		return selected != null;
	}
	
	/**
	 * Select the value for key for the selected database id, return null if
	 * no id had been selected or the key is not found
	 * 
	 * @param key the key to look up
	 * @return    the value of the key, if found, otherwise null
	 */
	public String findInRecord (String key) {
		return selected != null ? selected.get (key) : null;
	}

	/**
	 * Select the value for key for the selected database id, return dflt if
	 * no id had been selected or the key is not found
	 * 
	 * @param key  the key to look up
	 * @param dflt the default value to return
	 * @return     the value of the key, if found, otherwise dflt
	 */
	public String findInRecord (String key, String dflt) {
		String	rc = findInRecord (key);
		
		return rc != null ? rc : dflt;
	}
	public boolean findInRecord (String key, boolean dflt) {
		return Str.atob (findInRecord (key), dflt);
	}

	private static Pattern	parseLine = Pattern.compile ("([a-z0-9._+-]+):[ \t]*(.*)$", Pattern.CASE_INSENSITIVE | Pattern.UNICODE_CASE);
	private boolean readConfig () {
		if (content == null) {
			String	filename = configPath != null ? configPath : System.getenv ("DBCFG_PATH");
			File	fdesc;
			
			if (filename == null) {
				filename = Str.makePath ("$home", "etc", "dbcfg");
				fdesc = new File (filename);
				if (! fdesc.exists ()) {
					fdesc = new File (defaultConfigPath);
				}
			} else {
				fdesc = new File (filename);
			}
			try (FileInputStream stream = new FileInputStream (fdesc)) {
				byte[]	buffer = new byte[(int) fdesc.length ()];
				int	n = stream.read (buffer);
				
				if (n == buffer.length) {
					String	data = new String (buffer, "UTF8");
					content = new HashMap<>();
					
					for (String line : data.split ("\n")) {
						line = line.trim ();
						if ((line.length () == 0) || line.startsWith ("#")) {
							continue;
						}
						Matcher	m = parseLine.matcher (line);
						if (m.find ()) {
							String	id = m.group (1);
							String	tokens = m.group (2);
							HashMap	<String, String>
								entity = new HashMap<>();
							
							for (String token : tokens.split (", [ \t]*")) {
								String[]	parts = token.split ("[ \t]*=[ \t]*", 2);
								
								if (parts.length == 2) {
									entity.put (parts[0], parts[1]);
								}
							}
							content.put (id, entity);
						}
					}
				} else {
					lastError = "File " + filename + " incomplete, expected " + buffer.length + " bytes, but read " + n + " bytes";
				}
			} catch (FileNotFoundException e) {
				lastError = "File not found: " + filename;
			} catch (UnsupportedEncodingException e) {
				lastError = "File " + filename + " has invalid content (expecting UTF8): " + e.toString ();
			} catch (IOException e) {
				lastError = "File " + filename + " read error: " + e.toString ();
			}
		}
		return content != null;
	}
}
